### ISSUE TITLE
Browserify output is non-deterministic. Trying to cache-bust using hash of bundle contents.

### ISSUE THREAD

**Issue by mynameistechno** [browserify/browserify#1796]
I'm hashing the bundle contents for cache busting purposes but the bundle output changes even when there are no changes to the source files.

Sounds like this was previously addressed but regressed. See https://github.com/browserify/browserify/issues/711

Is there another way to do this? Thanks!

**Comment by z-vr** [comment]
same here 👋  i feel like it's because the order can be slightly different which can be seen in source maps 

```
-"lib/src/second.js",
"lib/src/first.js",
```
```
"lib/src/first.js",
+"lib/src/second.js",
```

**Comment by goto-bus-stop** [comment]
been hitting this too—i figured https://github.com/browserify/deps-sort should make sure this doesn't happen. no clue what's going on with that yet personally.

**Comment by casr** [comment]
I opened this issue up on Uglifyify. Perhaps it is somehow related to this? https://github.com/hughsk/uglifyify/issues/61#issuecomment-292025091

**Comment by abitrolly** [comment]
So is it fixed now? Is there a test?

**Comment by goto-bus-stop** [comment]
I don't believe anything related to this has landed recently. It's difficult to reproduce so I don't know what's causing it.

**Comment by danielcardoso5** [comment]
I'm also having this issue. The output files always have the same byte length but the imports order is different.

**Comment by goto-bus-stop** [comment]
Does anyone have a project where this happens regularly? I dont remember where I ran into it

**Comment by danielcardoso5** [comment]
I don't have anything set up with this issue that I can share.
I've set my cache bundle hash based on the file byte size and it's original name until there's a fix for this.
Not the best way, but it works for me right now.

**Comment by abitrolly** [comment]
Is it possible to check if imports order is unsorted?

**Comment by danielcardoso5** [comment]
I'm hashing my bundle names in the output function of split-require. I could reproduce it in the bundles even if I just added the split-require plugin without params.

**Comment by casr** [comment]
Hey @danielcardoso5, as goto-bus-stop mentioned, having a setup where this reproduces helps tremendously with testing. Perhaps as you have narrowed the cause in this case to when `split-require` is activated that you might be able to create a bare essentials test project. When I was testing the Uglifyify issue I made [this test project] which you can feel free to use as a base if it helps.

[this test project]: https://github.com/casr/uglifyify-bug

**Comment by danielcardoso5** [comment]
@casr ok, I'll set up a bare project where I can reproduce the issue.

**Comment by danielcardoso5** [comment]
I finally got time to setup a test project, here it is: [https://github.com/danielcardoso5/split-require-test](url). You have the steps to reproduce the issue in the readme. Let me know if you have any trouble running it.

**Comment by casr** [comment]
I ran the test that @danielcardoso5 put together (thanks btw!) and then diffed the bundles when they changed. I noticed on one occasion that the `split-require` bundles had only a difference in the source map. I decoded the source map and saw that in the `sources` key the files were listed in a different order.

Of course when a `split-require` bundle output changes, and the filename is based on a hash of the file, it will always change the hash on the main entry file because it references it. However, when diff-ing the entry bundle you can see the sources are moving around inside there as well.

I ran the test without defining `{ debug: true }` and it still had a problem so I do not think the issue lies with source maps.

@mynameistechno is your issue related to `split-require` too? @danielcardoso5's test definitely is and I think that issue should be moved over to the [`split-require` repo].

[`split-require` repo]: https://github.com/goto-bus-stop/split-require

**Comment by danielcardoso5** [comment]
@goto-bus-stop thanks, I can confirm the fix works!

@mynameistechno can you confirm this was also the issue you're having?

**Comment by mynameistechno** [comment]
> @danielcardoso5 are you using any transforms or plugins?

I am not using split-require ... Looks like I am using:

```
          'transform-flow-strip-types',
          'transform-class-properties',
          'transform-object-rest-spread'

  b.transform('babelify');
  b.transform('envify');

```

**Comment by casr** [comment]
@mynameistechno Based in a similar vein to the test before I tried to setup a test project based on the information you have shared. I just pushed it up on a branch https://github.com/casr/split-require-test/tree/babelify (ignore the project name; there is no split-require here!)

I couldn't get it to have a hash problem though so perhaps you can use drop in a failing test case?

**Comment by CanisLupus** [comment]
Hey guys, is this fixed by any chance? Mozilla Addons reviewers are rejecting my addon update because even though I provide scripts to reproduce the final package, the bundle generated by Browserify is not diff-accurate with my submission. It moves blocks around in the bundled file on their end... :(

**Comment by casr** [comment]
Are you using any transforms or plugins?

**Comment by casr** [comment]
@CanisLupus I found on your profile a Firefox extension called `swift-selection-search` but searching the repository for "browserify" only gave one result in issues. I believe you managed to package `iconv-lite` using Browseify but your build process was not committed to your `develop` branch (where it looks like your current activity is happening).

I would expect that you would need to at least pin a version in your `package.json` to have a stable build.

**Comment by CanisLupus** [comment]
Hey @casr, you won't find the original scripts on my repo. I use only the bundled version so that the addon doesn't need an extra dependency just to bundle the file.

You are right, the build scripts are not in the repo. The full recreation of the iconv-lite.js file needs two files:

generate.bat

	call npm init -y
	call npm install -g browserify@16.5.0
	call npm install iconv-lite@0.5.0
	call browserify index.js > iconv-lite.js

index.js

	window.iconv = require('iconv-lite');

This exact same process can (randomly?) put the contents of index.js `window.iconv = require('iconv-lite');` in line 2595 or in line 6508 of the resulting iconv-lite.js bundle (maybe other places?), also causing the indexes in blocks like `},{"safer-buffer":28}],25:[function(require,module,exports){` to change.

Am I doing anything wrong? I use browserify as-is; no plugins. The package.json is created by calling `npm init -y` and so has version 1.0.0.

**Comment by casr** [comment]
I put together [this gist](https://gist.github.com/casr/530c677bbec3e108d664445f132f644d) based on the above. Could you try the following and report back:

```sh
git clone git@gist.github.com:530c677bbec3e108d664445f132f644d.git determinism
cd determinism
npm install
node test.js
```

You should see something like this on success:
```
$ node test.js
..................................................
completed 50 successfully
```
and on failure the output might look more like:
```
...............
hash mismatch
6d95e4412a20ac26b3433dd0b783effdae800808 != adc83b19e793491b1c6ea0fd8b46cd9f32e592fc
failed after 17 runs
```

I currently only get the successful response so it might be a platform difference between us. Hopefully this might give us the next step.

**Comment by CanisLupus** [comment]
Thank you for taking the time to do this.

Unfortunately I haven't been able to reproduce the problem. It completed all 50 runs successfully. Twice. (Though I had to replace the first line with `var spawn = require('cross-spawn');` after getting some spawn errors, as per a solution I found on the web.)

I still haven't understood why it happens. I originally had the iconv-lite.js file with the `window.iconv = require('iconv-lite');` block probably 1/4 of the way through the file, but since I had to make it reproducible for Mozilla I re-did it weeks later and that block when to the very bottom this time, changing the indexes. Since it was consistently doing that each time I ran it, I submitted it this way, but the reviewer got that block in another place. Now, when I redo it myself I also get a different result.

Could this be affected by the current date? Feels like a dumb question, but I have no clue. :) I've tried changing dates in Windows and placing the files in different folders just to see if the path would do something, but no luck yet. Right now it's consistently outputting the same result, but I suspect it won't always.

**Comment by ljharb** [comment]
Are you sure you used the same node version? Do you have a lockfile, because if not, any number of dependencies could have updated and changed the result?

**Comment by casr** [comment]
> Are you sure you used the same node version? Do you have a lockfile, because if not, any number of dependencies could have updated and changed the result?

This is more likely going to be the problem when comparing your build to a recreated build at Mozilla.

But I think you might have implied that you had an issue locally? If so...

The test I put together made sure to use the local version of Browserify via NPM setting the correct `PATH`. Perhaps there is a chance your global Browserify has an issue.

Could you try change this in `test.js` and re-run?
```diff
-  var b = spawn('npm', ['run-script', '-s', 'build'])
+  var b = spawn('browserify', ['main.js'])
```

**Comment by CanisLupus** [comment]
> Are you sure you used the same node version?

Unfortunately I cannot be sure. In the instructions I send I tell them the version of npm used to create the submission (which is automatically extracted from `npm --version` and written in the README) but maybe they used another one. Either way, I'm using an old Node version and haven't updated (actually, maybe I should!), and saw varying results with that same version on the same machine.

> Do you have a lockfile, because if not, any number of dependencies could have updated and changed the result?

I'm not sure what you mean by lockfile. I'm not very familiar with Node and its packages, actually. I was assuming that if I install specific versions using this as seen above:

    call npm install -g browserify@16.5.0
    call npm install iconv-lite@0.5.0

I would get no trouble with dependencies updating. It is possible I'm doing something wrong, of course, but the full process uses only those 2 files I mentioned.

> But I think you might have implied that you had an issue locally? If so...

Yes.

> Could you try change this in `test.js` and re-run?

Unfortunately it's still consistently running all 50 times correctly.

Maybe it's best if I try updating everything I can and try to reproduce the problem again, since I feel that I am wasting your time without beeing sure of what the problem is. I really appreciate your help. If I find anything I'll let you know.

**Comment by casr** [comment]
The lockfile is useful for when your dependencies' dependencies need to be fixed to a particular version too. e.g. Browserify has a dependency on `deps-sort@^2.0.0` which means accept the latest version of `deps-sort` with major version 2. You can see how installing one day to the next might change the transitive dependency.

> Could this be affected by the current date?

Browserify collects up sources from the `require` calls. It then appears to sha them, dedupe and then sort them based on their hash. The package I mention above, `deps-sort`, actually shows an update from the end of October that changed the package that hashes them. Perhaps this change caused your issue?

**Comment by CanisLupus** [comment]
Thank you for explaining. So if deps-sort changes a non-major version, like going from something like 2.1.0 to 2.2.0, and I install Browserify 16.5.0 (a specific version that may not be the most recent one) it will still get the most recent version 2.x.x of deps-sort?

Pardon the novice question, but is this common in the world of JavaScript packages? I understand the _potential_ usefulness of getting automatic bug fixes on older versions of a package (in this case Browserify) if its dependencies update, but if those changes introduce new bugs it will also cause problems that are much harder to understand. Wouldn't it be better to depend on a very specific version? Feel free to not answer if this is too off-topic. :)

In my case, I assume this means that I shouldn't call `npm init -y` and should instead use a package.json and a package-lock.json that locks a specific version of deps-sort (and all other things) in place. And perhaps install Browserify non globally.

**Comment by casr** [comment]
I think NPM version 5 introduced the `package-lock.json` file (prior to that there were other mechanisms but I'll ignore that for now). With version 5, anytime you install you would get the lock file for your transitive dependencies. So let's work the Browserify example for `deps-sort`  (of course, this may not be the cause of the problem you experienced...) and assuming you have an NPM version of 5 or above.

```
$ npm view deps-sort time --json
{
  // ...
  "2.0.0": "2015-10-26T05:30:05.795Z",
  "2.0.1": "2019-10-29T12:35:44.053Z"
}
```

Above we can see `deps-sort`'s version history. As Browserify's `package.json` pins its version to `^2.0.0`, if you installed Browserify before 2019-10-29T12:35:44.053Z you would have gotten version 2.0.0 of `deps-sort`. This package would have been fixed in your `package-lock.json` file as well. Anyone installing with NPM after 2019-10-29T12:35:44.053Z using the `package.json` and `package-lock.json` generated before that date would still receive version 2.0.0 of `deps-sort` despite there being a newer version.

This is how you should be able to guarantee a stable build.

> In my case, I assume this means that I shouldn't call `npm init -y`

This is just a convenience to setup some common fields that NPM wants if you publish. The minimum valid `package.json` can be generated with simply `echo {} >package.json`.

> should instead use a package.json and a package-lock.json that locks a specific version of deps-sort (and all other things) in place. And perhaps install Browserify non globally.

Yes, this is a good reason to keep Browserify local to each project. For a project that is not intended to be a library for others, I would generally install dependencies with `--save-exact`/`-E` to maintain this build property. e.g. `npm install --save-dev --save-exact browserify` (can be shortened to `npm i -D -E browserify`).

We did kind of digress so hit me up on email if your curious about anything else. :) Should be easy to find if you go via my  GH profile.

**Comment by ljharb** [comment]
Yes, this is how semver works; deterministic builds are available by using a lockfile and aren’t valuable for many use cases.

Saving exact deps doesn’t give you any determinism (because of transitive deps); only a lockfile does, at which point exact deps add no value.

**Comment by casr** [comment]
> Saving exact deps doesn’t give you any determinism (because of transitive deps); only a lockfile does, at which point exact deps add no value.
Yes, you are right. Ignore the bit about `--save-exact`.

**Comment by CanisLupus** [comment]
Wow, thanks for the detailed explanation. :)

> This is how you should be able to guarantee a stable build.

All clear! So for the Mozilla submission my best bet is to provide a lockfile and the specific versions I use, forcing their build process to be the same, regardless of the range of versions Browserify accepts for its dependencies.

> We did kind of digress so hit me up on email if your curious about anything else. :)

I really appreciate your availability. :) I think I got it. And if I'm missing something I might read more about the way npm works online. Cheers!
